server/schedule/operator/operator.go

```diff
@@ -699,41 +699,128 @@ func CreateTransferLeaderOperator(desc string, region *core.RegionInfo, sourceSt
 		peer, err := cluster.AllocPeer(id)
 		if err != nil {
-			return nil, err
-		}
-		if cluster.IsRaftLearnerEnabled() {
-			steps = append(steps,
-				AddLearner{ToStore: id, PeerID: peer.Id},
-				PromoteLearner{ToStore: id, PeerID: peer.Id},
-			)
-		} else {
-			steps = append(steps, AddPeer{ToStore: id, PeerID: peer.Id})
+			log.Debug("peer alloc failed", zap.Error(err))
+			return kind, nil, err
 		}
+		addPeerSteps = append(addPeerSteps, CreateAddPeerSteps(cluster, id, peer.Id))
+		kind |= OpRegion
 	}
 
+	// Transferring leader to a new added follower may be refused by TiKV.
+	// Ref: https://github.com/tikv/tikv/issues/3819
+	// So, the new leader should be a follower that do not need to move if there is one,
+	// otherwise the first suitable new added follower.
+	orderedStoreIDs := append(oldStores, newStores...)
+
 	// Remove redundant peers.
+	var rmPeerSteps [][]OpStep
+	// Transfer leader as late as possible to prevent transferring to a new added follower.
+	var mvLeaderSteps []OpStep
 	for _, peer := range region.GetPeers() {
-		if _, ok := storeIDs[peer.GetStoreId()]; ok {
+		id := peer.GetStoreId()
+		if _, ok := targetStores[id]; ok {
 			continue
 		}
-		if region.GetLeader().GetId() == peer.GetId() {
-			for targetID := range storeIDs {
-				steps = append(steps, TransferLeader{FromStore: peer.GetStoreId(), ToStore: targetID})
-				break
+		if region.GetLeader().GetStoreId() == id {
+			tlkind, tlsteps, err := transferLeaderToSuitableSteps(cluster, id, orderedStoreIDs)
+			if err != nil {
+				log.Debug("move region to stores failed", zap.Uint64("region-id", region.GetID()), zap.Uint64s("store-ids", orderedStoreIDs), zap.Error(err))
+				return kind, nil, err
 			}
+			mvLeaderSteps = append(tlsteps, RemovePeer{FromStore: id})
+			kind |= tlkind
+		} else {
+			rmPeerSteps = append(rmPeerSteps, []OpStep{RemovePeer{FromStore: id}})
 		}
-		steps = append(steps, RemovePeer{FromStore: peer.GetStoreId()})
+		kind |= OpRegion
 	}
-	brief := fmt.Sprintf("mv region: stores %v to %v", u64Set(region.GetStoreIds()), u64Set(storeIDs))
-	return NewOperator(desc, brief, region.GetID(), region.GetRegionEpoch(), kind|OpRegion, steps...), nil
+
+	// Interleaving makes the operator add and remove peers one by one, so that there won't have
+	// too many additional peers if the operator fails in the half.
+	hint := len(addPeerSteps)*2 + len(rmPeerSteps) + len(mvLeaderSteps)
+	steps := interleaveStepGroups(addPeerSteps, rmPeerSteps, hint)
+
+	steps = append(steps, mvLeaderSteps...)
+
+	return kind, steps, nil
+}
+
+// interleaveStepGroups interleaves two slice of step groups. For example:
+//
+//  a = [[opA1, opA2], [opA3], [opA4, opA5, opA6]]
+//  b = [[opB1], [opB2], [opB3, opB4], [opB5, opB6]]
+//  c = interleaveStepGroups(a, b, 0)
+//  c == [opA1, opA2, opB1, opA3, opB2, opA4, opA5, opA6, opB3, opB4, opB5, opB6]
+//
+// sizeHint is a hint for the length of returned slice.
+func interleaveStepGroups(a, b [][]OpStep, sizeHint int) []OpStep {
```

Comment:
Why we need this hint?

Reply:
This hint is just like the second and third arguments in `make()`, for performance, not necessary.